File: fibonacci.js
// memoize
var fibs = {}

// fibonacci function
function fib(n) {
  if (n <= 1) return n
  if (fibs[n]) return fibs[n]
  return fibs[n] = fib(n-1) + fib(n-2)
}

for (let i = 0; i < 50; i++) {
  console.log(fib(i))
}

File: fibonacci.kt
val fibs = mutableMapOf(0 to 0L, 1 to 1L)

fun main(args: Array<String>) {
  for (i in 0..50) {
    println(fib(i))
  }
}

// fibonacci function
fun fib(n: Int): Long? {
  return if (fibs.containsKey(n)) fibs[n]
        else {
          fibs.put(n, fib(n-1)!!+fib(n-2)!!)
          fibs[n]
        }
}